### PR TITLE
Implemented KRYPTO.ecdsaRecover hook

### PR DESCRIPTION
**Pull request by u--**
Duplicate PR because the old one got closed accidentally.
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 1]
```diff
@@ -97,3 +111,147 @@ evalKeccak =
...
+        , i2bs x
+        , i2bs y
+        ]
+encodePoint _ _ = error "Should never obtain point-at-infinity here!"
+
+bs2s :: ByteString -> String
+bs2s = map (chr . fromIntegral) . ByteString.unpack
+
+s2i :: String -> Integer
+s2i = bs2i . ByteString.pack . map (fromIntegral . ord)
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: Is this going to fail with a cryptic error message if the `String`'s code points are outside the 8-bit range?

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 2]
```diff
@@ -97,3 +111,147 @@ evalKeccak =
...
+        (if even y then 0x02 else 0x03)
+        (i2bs x)
+    | otherwise = ByteString.concat
+        [ ByteString.pack [0x04]
+        , i2bs x
+        , i2bs y
+        ]
+encodePoint _ _ = error "Should never obtain point-at-infinity here!"
+
+bs2s :: ByteString -> String
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: Please give these four utility functions meaningful names. I can infer from context that they are to convert between `ByteString`, `Integer`, and `String`, but what that "conversion" entails is not immediately evident.

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 3]
```diff
@@ -97,3 +111,147 @@ evalKeccak =
...
+bs2i =
+    ByteString.foldr (\word i -> fromIntegral word + (i `shiftL` 8)) 0
+  . ByteString.reverse
+
+i2bs :: Integer -> ByteString
+i2bs = ByteString.reverse .
+    ByteString.unfoldr
+    (\i -> if i == 0
+        then Nothing
+        else Just (fromIntegral (i `mod` 256), i `div` 256)
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: Consider using `divMod` here.

File: src/main/haskell/kore/test/Test/Kore/Builtin/Krypto.hs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE MagicHash #-}
```
Comment (ttuegel) [inline src/main/haskell/kore/test/Test/Kore/Builtin/Krypto.hs]: I don't think this module needs this language extension.

File: src/main/haskell/kore/test/Test/Kore/Builtin/Krypto.hs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+--- s = "\x14\x8c\x87*7X\xdb0\xe6\xcf\xf6=_\x8c\x11U\xbd\xf1=\x9a9\xbf\x99;B\xd3\xf2\xccz\fV\x8a"
+--- result = ":QAvFo\xa8\x15\xedH\x1f\xfa\xd0\x91\x10\xa2\xd3D\xf6\xc9\xb7\x8c\x1d\x14\xaf\xc3Q\xc3\xa5\x1b\xe3=\x80r\xe7y9\xdc\x03\xbaDy\x07y\xb7\xa1\x02[\xaf0\x03\xf6s$0\xe2\f\xd9\xb7m\x953\x91\xb3"
+
+-- messageHash = "!\\\x9f\x84\x95R\x9d\x10\xd1[h\xb2\x9aJ\xf2\xbe\x01\xda\xebeK\x0e.\xc2T_\xc0c\xa0\xe2\x89O"
+-- v = 28
+-- r = "\xc6R \xd2.\xe9S>i\xca\f\xcaO-\x90\xdb\xab\xd0\xdbq\xe8h\x83\xdd\xa7\x9aa5\xa8\x96\xc6\xc2"
+-- s = "\x14\x8c\x87*7X\xdb0\xe6\xcf\xf6=_\x8c\x11U\xbd\xf1=\x9a9\xbf\x99;B\xd3\xf2\xccz\fV\x8a"
+
+result :: ByteString
+result = ByteString.pack $ map (fromIntegral . ord)
```
Comment (ttuegel) [inline src/main/haskell/kore/test/Test/Kore/Builtin/Krypto.hs]: Could you export one of the conversion functions you wrote for this? If it is useful here, other consumers of our library would probably find it useful.

File: src/main/haskell/kore/src/Kore/Builtin/Builtin.hs [hunk 6]
```diff
@@ -611,17 +612,17 @@ ternaryOperator
             [_, _, _] -> return NotApplicable
             _ -> wrongArity (Text.unpack ctx)
 
-functionEvaluator
-    ::  (forall variable
+type FunctionImplementation
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Builtin.hs]: :+1:

File: src/main/haskell/kore/src/Kore/Builtin/Builtin.hs [hunk 7]
```diff
@@ -611,17 +612,17 @@ ternaryOperator
...
         .  (Ord (variable Object), Show (variable Object))
         => MetadataTools Object StepperAttributes
         -> StepPatternSimplifier Object variable
         -> Sort Object
         -> [StepPattern Object variable]
         -> Simplifier (AttemptedFunction Object variable)
-        )
-    -- ^ Builtin function implementation
-    -> Function
+
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Builtin.hs]:
```suggestion
```

File: src/main/haskell/kore/src/Kore/Builtin/Builtin.hs [hunk 8]
```diff
@@ -611,17 +612,17 @@ ternaryOperator
             [_, _, _] -> return NotApplicable
             _ -> wrongArity (Text.unpack ctx)
 
-functionEvaluator
-    ::  (forall variable
+type FunctionImplementation
+  = forall variable
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Builtin.hs]: Please check the indentation here against the style guide.

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 9]
```diff
@@ -32,26 +44,24 @@ import           Data.String
...
-import           Kore.Step.StepperAttributes
-                 ( StepperAttributes )
+
+import Control.Exception.Base
+       ( assert )
 
 keccakKey :: IsString s => s
 keccakKey = "KRYPTO.keccak256"
 
+ecsdaRecover :: String
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: May I suggest,
```suggestion
ecdsaRecover :: IsString s => s
```
You will need to `import Data.String (IsString)` above, and you may delete the other definition.

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 10]
```diff
@@ -32,26 +44,24 @@ import           Data.String
...
-                 ( Sort )
-import           Kore.Step.Function.Data
-                 ( AttemptedFunction )
-import           Kore.Step.Pattern
-import           Kore.Step.Simplification.Data
-                 ( Simplifier, StepPatternSimplifier )
-import           Kore.Step.StepperAttributes
-                 ( StepperAttributes )
+
+import Control.Exception.Base
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: This needs to be moved up with the other standard library imports.

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 11]
```diff
@@ -32,26 +44,24 @@ import           Data.String
 import           Data.Text
                  ( Text )
 import qualified Data.Text.Encoding as Text
-
-import           Kore.AST.MetaOrObject
-                 ( Object )
+import           Data.Word
+                 ( Word8 )
 import qualified Kore.Builtin.Builtin as Builtin
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: Please separate these blocks of imports following the style guide.

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 12]
```diff
@@ -97,3 +111,147 @@ evalKeccak =
                 result = fromString (show digest)
             Builtin.appliedFunction
                 $ String.asExpandedPattern resultSort result
+
+evalECDSARecover :: Builtin.Function
+evalECDSARecover =
+    Builtin.functionEvaluator eval0
+
+eval0 :: Builtin.FunctionImplementation
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: Please either move this implementation to a `where` clause, or give it a meaningful name. It is not immediately obvious that it has anything to do with `evalECDSARecover`.

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 13]
```diff
@@ -97,3 +111,147 @@ evalKeccak =
...
+
+signatureToKey
+    :: Integer
+    -> Integer
+    -> Integer
+    -> Integer
+    -> ByteString
+signatureToKey messageHash r s v =
+    assert (28 <= v && v <= 34)
+  $ ByteString.drop 1
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: Please check the style guide re: indentation here.

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 14]
```diff
@@ -97,3 +111,147 @@ evalKeccak =
...
+    recId = v - 27
+    compressed = v >= 31
+
+recoverPublicKey
+    :: Integer
+    -> (Integer, Integer)
+    -> Integer
+    -> Point
+recoverPublicKey recId (r, s) e =
+    assert (recId > 0)
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: Please check the style guide re: indentation here.

File: src/main/haskell/kore/src/Kore/Builtin/Krypto.hs [hunk 15]
```diff
@@ -97,3 +111,147 @@ evalKeccak =
...
+-- Leading bit signals whether the point is compressed.
+-- Superfluous because we drop it later on.
+-- Kept here for completeness sake, to match
+-- the code in the java backend.
+encodePoint
+    :: Bool
+    -> Point
+    -> ByteString
+encodePoint compressed (Point x y)
+    | compressed =
```
Comment (ttuegel) [inline src/main/haskell/kore/src/Kore/Builtin/Krypto.hs]: Please check the style guide re: indentation here.

File: src/main/haskell/kore/test/Test/Kore/Builtin/Krypto.hs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+import Hedgehog
+import Test.Tasty
+import Test.Tasty.Hedgehog
+
+import           Data.ByteString
+                 ( ByteString )
+import qualified Data.ByteString as ByteString
+import           Data.Char
+--- messageHash = "!\\\x9f\x84\x95R\x9d\x10\xd1[h\xb2\x9aJ\xf2\xbe\x01\xda\xebeK\x0e.\xc2T_\xc0c\xa0\xe2\x89O"
```
Comment (ttuegel) [inline src/main/haskell/kore/test/Test/Kore/Builtin/Krypto.hs]: Could we remove the dead code here?

File: src/main/haskell/kore/test/Test/Kore/Builtin/Krypto.hs (hunk 4 above)
Comment (u--) [reply src/main/haskell/kore/test/Test/Kore/Builtin/Krypto.hs]: Weird, I don't think I added that, must've been from an old commit

**Comment by ttuegel** [comment]
Jenkins: Add to whitelist

**Pull request merged**